demos/fleet_provisioning/fleet_provisioning_with_csr/fleet_provisioning_csr_demo.c

```diff
@@ -0,0 +1,874 @@
+    /* Buffer for holding the certificate ID. */
+    char certificateId[ CERT_ID_BUFFER_LENGTH ];
+    size_t certificateIdLength;
+    /* Buffer for holding the certificate ownership token. */
+    char ownershipToken[ OWNERSHIP_TOKEN_BUFFER_LENGTH ];
+    size_t ownershipTokenLength;
+    bool connectionEstablished = false;
+    size_t i;
+    int demoRunCount = 0;
+
+    /* Silence compiler warnings about unused variables. */
+    ( void ) argc;
+    ( void ) argv;
+
+    do
+    {
+        /* Initialize the buffer lengths to their max lengths. */
+        certificateLength = CERT_BUFFER_LENGTH;
+        certificateIdLength = CERT_ID_BUFFER_LENGTH;
+        ownershipTokenLength = OWNERSHIP_TOKEN_BUFFER_LENGTH;
+
+        /* The demo template we use has a serial number as a parameter.
+         * For this demo we use a random 10 digit number. */
+        for( i = 0; i < serialLength; i++ )
+        {
+            serial[ i ] = '0' + ( rand() % 10 );
+        }
+
+        /* Read the CSR into the CSR buffer. */
+        status = getCsr( csr, NETWORK_BUFFER_SIZE, &csrLength );
+
+        if( status == true )
+        {
+            /* Create the document containing the CSR to publish to the
+             * CreateCertificateFromCsr API. */
+            status = generateCsrRequest( payloadBuffer,
+                                         NETWORK_BUFFER_SIZE,
+                                         csr,
+                                         csrLength,
+                                         &payloadLength );
+        }
+
+        /**** Connect to AWS IoT Core with provisioning claim credentials *****/
+
+        /* We first use the claim credentials to connect to the broker. These
+         * credentials should allow use of the RegisterThing API and one of the
+         * CreateCertificatefromCsr or CreateKeysAndCertificate.
+         * In this demo we use CreateCertificatefromCsr. */
+        if( status == true )
+        {
+            /* Attempts to connect to the AWS IoT MQTT broker over TCP. If the
+             * connection fails, retries after a timeout. Timeout value will
+             * exponentially increase until maximum attempts are reached. */
+            LogInfo( ( "Establishing MQTT session with claim certificate..." ) );
+            status = EstablishMqttSession( provisioningPublishCallback,
+                                           CLAIM_CERT_PATH,
+                                           CLAIM_PRIVATE_KEY_PATH );
+
+            if( status == false )
+            {
+                LogError( ( "Failed to establish MQTT session." ) );
+            }
+            else
+            {
+                connectionEstablished = true;
+            }
+        }
+
+        /**** Call the CreateCertificateFromCsr API ***************************/
+
+        /* We use the CreateCertificatefromCsr API to obtain a client certificate
+         * for a key on the device by means of sending a Certificate Signing
```

Comment:
```suggestion
         * for a private key on the device by means of sending a Certificate Signing
```

Reply:
There can be some comments to suggest customers that CreateCertificateFromCsr is recommended over CreateKeysAndCertificate to keep the private keys within the device

Reply:
These doesn't seem to be resolved